Authored TAA and CAS docs (#793)
o3de/o3de.org: content/docs/atom-guide/features/taa.md

---

> title: Temporal Anti-Aliasing

Reviewer: ```suggestion
title: Temporal Anti-aliasing
```
When capitalizing "anti-aliasing", even in a title, only "Anti-" is capitalized.

> description: "Learn about temporal anti-aliasing in the Atom Renderer"

Reviewer: ```suggestion
description: "Learn about temporal anti-aliasing (TAA) in Atom Renderer, the graphics engine integrated into Open 3D Engine (O3DE)."
```
* Added the acronym "TAS".
* Consider mentioning O3DE in the description for SEO integration. I added "the graphics engine integrated into [O3DE]" as an initial suggestion for you to consider.

toc: true
---  

*Temporal anti-aliasing* (TAA) is a type of supersampling that takes samples from different locations within a pixel each frame and combines them with samples from previous frames to create a supersampled image. The main challenge of TAA is knowing when samples from the previous frame are valid. TAA may cause some softness in your image, especially while in motion. To improve the quality of your image, you can enable contrast adaptive sharpening, which sharpens the details of your image. For more information, refer to [Contrast Adaptive Sharpening](cas/).

## Parameters

| Parameter | Description | Values | Default |
| - | - | - | - |
| **m_currentFrameContribution** | Controls how much the current frame's samples contribute to the final image, when history isn't rejected. Reducing this value may help prevent flickering in some circumstances. | `0.1` to `1.0` | `0.1` |
| **m_clampGamma** | Controls the number of standard deviations from the mean of the 3x3 current neighborhood that the history is clipped to. Increasing this value may prevent flickering, but can also lead to ghosting. | `0.0` to infinity | `1.0` |

> | **m_maxDeviationBeforeDampening** | Controls the number of standard deviations from the mean, which defines a threshold for the pixel's brightness before it begins to reduce its contribution. This prevents flickering by lowering the weight of the current pixel, if its brightness is very different from the surrounding neighborhood and history. By default, any pixel with a brightness that is more than 0.5 standard deviations from the mean will begin to contribute less. | `0.0` to infinity | `0.5` |

Reviewer: ```suggestion
| **m_maxDeviationBeforeDampening** | Controls the number of standard deviations from the mean, which defines a threshold for the pixel's brightness before it begins to reduce its contribution. This prevents flickering by lowering the weight of the current pixel, if its brightness is very different from the surrounding neighborhood and history. By default, any pixel with a brightness that is more than 0.5 standard deviations from the mean begins to contribute less. | `0.0` to infinity | `0.5` |
```
Minor change from future tense to present tense (preferred).

## Using TAA

You can enable `Taa.pass` in the main render pipeline through the parent pass, `PostProcessParent.pass`. By default, TAA is disabled. To enable it, set `Enabled` to `true` in the parent pass.

You can also add `Taa.pass` to a custom pipeline.

## Issues and improvements

> The following are a list of known issues and improvements for TAA:

Reviewer: ```suggestion
The following is a list of known issues and improvements for TAA:
```

> The following is a list of known issues and improvements for TAA:

Reviewer 2: For everything below this line, if it is truly an issue, it should be in the known issues list of release notes. If it is a tip or trick, it should be a call-out of some sort, either on this reference page or in a tutorial. Non-specific claims for future features and improvements should not be included in documentation.

- TAA can cause Bloom to produce flickering in the resulting image.

> - TAA can cause Aux-Geom to produce jittering in the resulting image. The reason is that Aux-Geom uses the jittered view but applies it after TAA runs. Therefore, TAA is recommended to run in Game mode, rather than Edit mode.

Reviewer: ```suggestion
- TAA can cause Aux-Geom to produce jittering in the resulting image. The reason is that Aux-Geom uses the jittered view but applies it after TAA runs. Therefore, we recommend running TAA in Game mode rather than in Edit mode.
```
Passive voice fix

> - TAA involves motion vectors, which when used on parallel occlusion mapping (POM) can cause blur in the resulting image. The reason is that motion vectors become more inaccurate the greater the depth offset is from POM. You can enable pixel depth offset to reduce the blur, but it can impact performance.

Reviewer: ```suggestion
- TAA involves motion vectors, which when used on parallel occlusion mapping (POM) can cause blur in the resulting image. The reason is that motion vectors become more inaccurate the greater the depth offset is from POM. To reduce the blur, you can enable pixel depth offset, but it can impact performance.
```
I recommend leading with the intent/infinitive phrase ("To reduce the blur") in the last sentence here.

> - Additionally, transparencies don't write to motion vectors so transparent objects, such as particles, may have increased blur.

Reviewer: ```suggestion
- Transparencies don't write to motion vectors, so transparent objects, such as particles, may have increased blur.
```
Using "additionally" makes this line follow directly from the bullet point before it. I see the connection (talking about motion vectors and blur), but it looks weird when these bullets are at the same level. I recommend either dropping the conjunction as shown here, or either (a) moving this line up as another sentence in the previous list item, or (b) indenting this list item to more directly show how it follows.

>   - Additionally, transparencies don't write to motion vectors so transparent objects, such as particles, may have increased blur.

Reviewer: ```suggestion
  - Additionally, transparencies don't write to motion vectors, so transparent objects, such as particles, may have increased blur.
```
I still recommend adding a comma after "motion vectors" for flow. ("so transparent objects" is actually the start of a new clause with a new subject.)

- TAA can cause screen space ambient occlusion (SSAO) to produce flickering in the resulting image, especially when SSAO runs at half resolution. The reason is that TAA involves jittering the camera to collect samples in different parts of a pixel in each frame.

> - In the future, the **O3DE Editor** will support a TAA component, making it easier to configure TAA in your scene.

Reviewer: ```suggestion
- In the future, **O3DE Editor** will support a TAA component, making it easier to configure TAA in your scene.
```

| **m_maxDeviationBeforeDampening** | Controls the number of standard deviations from the mean, which defines a threshold for the pixel's brightness before it begins to reduce its contribution. This prevents flickering by lowering the weight of the current pixel, if its brightness is very different from the surrounding neighborhood and history. By default, any pixel with a brightness that is more than 0.5 standard deviations from the mean will begin to reduce its contribution. | `0.5` |

## Using TAA

You can enable `Taa.pass` in the main render pipeline through the parent pass, `PostProcessParent.pass`. By default, TAA is disabled. To enable it, set `Enabled` to `true` in the parent pass.

You can also add `Taa.pass` to a custom pipeline. 

## Issues and improvements

The following are a list of known issues and improvements for TAA:

- The bloom effect causes flickering when TAA is enabled. 
  
- Aux-Geom causes jittering because it uses the jittered view, but applies it after TAA runs. Therefore, TAA currently is not recommended to use in the **O3DE Editor**.

> - Motion vectors on parallel occlusion mapping are only accurate when pixel depth offset is enabled. However, this can lead to additional blur when in motion.

Reviewer 3: The way this is worded is a little backwards. Motion vectors on POM won't be accurate unless PDO is enabled - the greater the depth offset from pom, the more inaccurate the motion vectors may become. Inaccuracies in the motion vectors is what can lead to additional blur (only when the object or camera is in motion though). If PDO is enabled this problem goes away, but PDO is expensive so that's not really recommended. Usually standard history rejection and rectification make this problem hard to notice, but it depends a lot on the material and mesh, so ymmv.

- Transparent objects, such as particles, may have increased blur, because transparencies don't write motion vectors.

> - Screen space ambient occlusion (SSAO) leads to camera jitter, which can cause flickering if TAA is using half-sized render targets. To mitigate this issue, SSAO can be increased to full resolution.

Reviewer 3: So TAA jitters the camera (not ssao), it's part of how the effect works to grab samples in different parts of a pixel each frame. This jittering can cause noise in SSAO though, especially when SSAO is run at half res (which is its default).

> - To improve the quality of TAA, mip bias should be added to all shaders. Since TAA is a supersampling method, it's best to sample from a higher mip level.

Reviewer 3: This isn't something we're asking customers to do, it's another thing we need to do on our end but haven't yet. I'm not sure if this is even really worth mentioning because it's not really a problem as much as it is a way that we'll be able to make TAA look even better in the future.

> - In the future, the O3DE Editor will support a TAA component, making it easier to configure TAA in your scene.

Reviewer 3: Same goes for CAS
